test/Microsoft.Azure.Devices.Edge.Test/EndToEnd.cs

```diff
@@ -0,0 +1,160 @@
+                name);
+        }
+
+        [Test]
+        public async Task ModuleToModuleDirectMethod(
+            [Values("Amqp", "Mqtt")] string edgeToCloud,
+            [Values("Amqp", "Mqtt")] string moduleToEdge)
+        {
+            string edgeAgent = Context.Current.EdgeAgent.Expect(() => new ArgumentException());
+            string edgeHub = Context.Current.EdgeHub.Expect(() => new ArgumentException());
+            string senderImage = Context.Current.MethodSender.Expect(() => new ArgumentException());
+            string receiverImage = Context.Current.MethodReceiver.Expect(() => new ArgumentException());
+            Option<Uri> proxy = Context.Current.Proxy;
+
+            CancellationToken token = this.cts.Token;
+
+            string name = $"module-to-module direct method (upstream:{edgeToCloud}, modules:{moduleToEdge})";
+            Log.Information("Running test '{Name}'", name);
+            await Profiler.Run(
+                async () =>
+                {
+                    var iotHub = new IotHub(
+                        Context.Current.ConnectionString,
+                        Context.Current.EventHubEndpoint,
+                        proxy);
+
+                    EdgeDevice device = (await EdgeDevice.GetIdentityAsync(
+                        Context.Current.DeviceId,
+                        iotHub,
+                        token)).Expect(() => new Exception("Device should have already been created in setup fixture"));
+
+                    string methodSender = $"methodSender-{edgeToCloud}-{moduleToEdge}";
+                    string methodReceiver = $"methodReceiver-{edgeToCloud}-{moduleToEdge}";
+
+                    var builder = new EdgeConfigBuilder(device.Id);
+                    Context.Current.Registry.ForEach(
+                        r => builder.AddRegistryCredentials(r.address, r.username, r.password));
+                    builder.AddEdgeAgent(edgeAgent)
+                        .WithEnvironment(new[] { ("UpstreamProtocol", edgeToCloud) })
+                        .WithProxy(proxy, Enum.Parse<Protocol>(edgeToCloud));
+                    builder.AddEdgeHub(edgeHub)
+                        .WithEnvironment(new[] { ("UpstreamProtocol", edgeToCloud) })
+                        .WithProxy(proxy, Enum.Parse<Protocol>(edgeToCloud));
+                    builder.AddModule(methodSender, senderImage)
+                        .WithEnvironment(
+                            new[]
+                            {
+                                ("UpstreamProtocol", moduleToEdge),
+                                ("TargetModuleId", methodReceiver)
+                            });
+                    builder.AddModule(methodReceiver, receiverImage)
+                        .WithEnvironment(new[] { ("UpstreamProtocol", moduleToEdge) });
+                    await builder.Build().DeployAsync(iotHub, token);
+
+                    var hub = new EdgeModule("edgeHub", device.Id, iotHub);
```

Comment:
should we make "edgeHub" as a constant or add a new class EdgeHubModule extended from EdgeModule and only takes device id and iotHub as constructor argument?

Reply:
I don't see much use for that yet. We just need to wait for edge hub to enter the `running` state along with the other modules, and `EdgeModule` gives us that ability. The only reason I created a special derived class for `EdgeAgent` was to add the `PingAsync` method.